op-node/node/safedb/safedb.go

```diff
@@ -0,0 +1,192 @@
+		return fmt.Errorf("failed to record safe head update: %w", err)
+	}
+	if err := batch.Commit(d.writeOpts); err != nil {
+		return fmt.Errorf("failed to commit safe head update: %w", err)
+	}
+	return nil
+}
+
+func (d *SafeDB) SafeHeadReset(safeHead eth.L2BlockRef) error {
+	d.m.Lock()
+	defer d.m.Unlock()
+	iter, err := d.db.NewIter(SafeByL1BlockNumKey.IterRange())
+	if err != nil {
+		return fmt.Errorf("reset failed to create iterator: %w", err)
+	}
+	defer iter.Close()
+	if valid := iter.SeekGE(SafeByL1BlockNumKey.Of(safeHead.L1Origin.Number)); !valid {
+		// Reached end of column without finding any entries to delete
+		return nil
+	}
+	for {
+		val, err := iter.ValueAndErr()
+		if err != nil {
+			return fmt.Errorf("reset failed to read entry: %w", err)
+		}
+		l1Block, l2Block, err := DecodeSafeByL1BlockNum(iter.Key(), val)
+		if err != nil {
+			return fmt.Errorf("reset encountered invalid entry: %w", err)
+		}
+		if l2Block.Number >= safeHead.Number {
+			// Keep a copy of this key - it may be modified when calling Prev()
+			l1HeadKey := slices.Clone(iter.Key())
+			hasPrevEntry := iter.Prev()
+			// Found the first entry that made the new safe head safe.
+			batch := d.db.NewBatch()
+			if err := batch.DeleteRange(l1HeadKey, SafeByL1BlockNumKey.Max(), d.writeOpts); err != nil {
+				return fmt.Errorf("reset failed to delete entries after %v: %w", l1HeadKey, err)
+			}
+
+			// If we reset to a safe head before the first entry, we don't know if the new safe head actually became
+			// safe in that L1 block or if it was just before our records start, so don't record it as safe at the
+			// specified L1 block.
+			if hasPrevEntry {
+				if err := batch.Set(l1HeadKey, SafeByL1BlockNumValue(l1Block, safeHead.ID()), d.writeOpts); err != nil {
+					return fmt.Errorf("reset failed to record safe head update: %w", err)
+				}
+			}
+			if err := batch.Commit(d.writeOpts); err != nil {
+				return fmt.Errorf("reset failed to commit batch: %w", err)
+			}
+			return nil
+		}
+		if valid := iter.Next(); !valid {
+			// Reached end of column without finding any entries to delete
+			return nil
+		}
+	}
+}
+
+func (d *SafeDB) SafeHeadAtL1(ctx context.Context, l1BlockNum uint64) (l1Block eth.BlockID, safeHead eth.BlockID, err error) {
+	d.m.RLock()
+	defer d.m.RUnlock()
+	iter, err := d.db.NewIterWithContext(ctx, SafeByL1BlockNumKey.IterRange())
+	if err != nil {
+		return
+	}
+	defer iter.Close()
+	if valid := iter.SeekLT(SafeByL1BlockNumKey.Of(l1BlockNum + 1)); !valid {
+		err = ErrNotFound
+		return
+	}
+	// Found an entry at or before the requested L1 block
+	val, err := iter.ValueAndErr()
+	if err != nil {
+		return
+	}
+	l1Block, safeHead, err = DecodeSafeByL1BlockNum(iter.Key(), val)
+	return
+}
+
+func (d *SafeDB) Close() error {
+	d.m.Lock()
+	defer d.m.Unlock()
+	if d.closed {
+		// Already closed
+		return nil
+	}
+	d.closed = true
+	return d.db.Close()
```

Comment:
If it fails to close, do we want to be able to retry closing with a second call to `SafeDB.Close`?

Reply:
hmm, I'm not sure that's safe - pebbledb explicitly says not to call `Close` multiple times. It's unclear if that also applies if it returned an error though.

Yeah the first thing the pebbledb `Close` does is:
```
	if err := d.closed.Load(); err != nil {
		panic(err)
	}

	// Clear the finalizer that is used to check that an unreferenced DB has been
	// closed. We're closing the DB here, so the check performed by that
	// finalizer isn't necessary.
	//
	// Note: this is a no-op if invariants are disabled or race is enabled.
	invariants.SetFinalizer(d.closed, nil)

	d.closed.Store(errors.WithStack(ErrClosed))
	close(d.closedCh)
```

so if you call `Close` twice it will panic even if it failed the first time.